src/analyses/ai.cpp

```diff
@@ -304,231 +308,223 @@ bool ai_baset::visit(
 
 bool ai_baset::visit_edge(
   const irep_idt &function_id,
-  locationt l,
+  trace_ptrt p,
   const irep_idt &to_function_id,
-  const locationt &to_l,
+  locationt to_l,
   const namespacet &ns,
   working_sett &working_set)
 {
+  // Has history taught us not to step here...
+  auto next = p->step(to_l, *(storage->abstract_traces_before(to_l)));
+  if(next.first == ai_history_baset::step_statust::BLOCKED)
```

Comment:
At the moment I was using `std::pair` because I figured it was better to use the standard types than to replicate stuff but if you want I can create a `step_return_valuet` object?